[admin] Added JS widget for notififcations

pandafy opened the pull request:
TODOS

- [x] Show loader when request to server is made.
- [x] Setup ~JSLint~ JSHint 
- [ ] Add feature for #9 

Closes #4 

![Hnet-image (9)](https://user-images.githubusercontent.com/32094433/85473873-7f7e5700-b5d1-11ea-90c1-30793466e61c.gif)

pandafy commented:
Thanks @nemesisdesign , I will make these changes in the UI. 

> How are you flagging notifications as read now?

Through AJAX calls to API, there is an endpoint to flag a single notification as read. 
If you are asking about visually, read notifications have a faded background color.. 

> Is the bell icon updating the count?

It will get updated after a page reload for now. After implementing web sockets we can update it instantly. 

I'll get this PR to an acceptable level and will start working on web sockets simultaneously. 😄

Vivekrajput20 commented on openwisp_notifications/static/openwisp_notifications/js/notifications.js:
```diff
@@ -5,9 +5,12 @@ django.jQuery(function($) {
         new_desc;
     window.gettext = window.gettext || function (t) { return t };
     if (data.url) {
+        print("ASdasd")
```
remove this line

Vivekrajput20 commented:
@pandafy I think you should setup eslint instead of JSLint. Eslint is the most recent and is highly extensible. It also has a plugin for django [eslint-plugin-django](https://www.npmjs.com/package/eslint-plugin-django)
Also we have already integrated it in some of our repositories like [openwisp-wifi-login-pages](https://github.com/openwisp/openwisp-wifi-login-pages), [netjsongraph.js](https://github.com/openwisp/netjsongraph.js). So it will be easier for you to setup it.

Vivekrajput20 commented:
Regarding the web sockets implementation,  are you planning to use [django-channels](https://channels.readthedocs.io/en/latest/) @pandafy ?
@nemesisdesign  do we use [gunicorn](https://gunicorn.org/) server for deploying it on production or any other server?

pandafy commented:
> Regarding the web sockets implementation, are you planning to use [django-channels](https://channels.readthedocs.io/en/latest/) @pandafy ?

Yes @Vivekrajput20 .

pandafy commented:
@Vivekrajput20 I have configured JSHint to keep JS linting simple, what do you say?  Should I move to ESLint?

Vivekrajput20 commented on openwisp_notifications/templatetags/notification_tags.py:
```diff
@@ -32,4 +32,23 @@ def unread_notifications(context):
+        '''
+        <div class="notification-dropdown hide">
+            <div class="filters">
+                <span class="btn" id="mark-all-read">Mark all read</span>
```
disable onClick event when these buttons are disabled. Currently theses buttons send API request onClick even if they are disabled.
Also either use `<button>` tag or add role and tabindex attributes to them

Vivekrajput20 commented on openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:
```diff
@@ -0,0 +1,99 @@
+}
+
+.accordion {
+  overflow-y: scroll;
```
make it auto

Vivekrajput20 commented on openwisp_notifications/templatetags/notification_tags.py:
```diff
@@ -32,4 +32,23 @@ def unread_notifications(context):
+            </div>
+            <div class="accordion"></div>
+            <div class="no-notifications hide">
+                <p>No notifications founds.</p>
```
Change this text to "No new notification"

Vivekrajput20 commented on openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:
```diff
@@ -0,0 +1,99 @@
+}
+
+.accordion-panel {
+  padding: 0 18px;
```
add some top padding

Vivekrajput20 commented:
good work @pandafy 
these are some of the things I noticed and needs to be added/fixed

- Infinite scroll does not work until the user clicks on any of the notification. 
- mark all read and mark read does not update the notification count on the bell icon
- If the last notification is clicked the content is opened but content is out of the window and not visible
- Current background colors for read and unread notification are not very intuitive, I think we can keep white background for read notifications and give background color to unread notifications only. What do you think @pandafy  @nemesisdesign?

Vivekrajput20 commented on openwisp_notifications/static/openwisp_notifications/css/notifications.css:
```diff
@@ -19,10 +19,10 @@
+  height: 14px;
+  vertical-align: middle;
+  position: relative;
+  top: -1px;
```
add `cursor:pointer`

Vivekrajput20 commented on openwisp_notifications/static/openwisp_notifications/css/notification_widget.css:
```diff
@@ -0,0 +1,99 @@
+  text-transform: none;
+  background-color: whitesmoke;
+  border-radius: 5px;
+  box-shadow: 0px 1px 20px 3px black;
```
reduce the blur and spread in box-shadow

nemesisdesign commented on run-qa-checks:
```diff
@@ -1,5 +1,8 @@
 set -e
+
+jshint ./openwisp_notifications/static/openwisp_notifications/js/*.js
+
```
can we also add linting for the CSS as we have in https://github.com/openwisp/OpenWISP-Website/blob/master/run-qa-checks#L6?

pandafy commented:
> * Infinite scroll does not work until the user clicks on any of the notification.

I am not able to reproduce this. :sweat_smile: 

> * mark all read and mark read does not update the notification count on the bell icon

This was because web sockets were not added on this PR. 

> * If the last notification is clicked the content is opened but content is out of the window and not visible
> * Clicking inside the notification dropdown header leads to closing of the notification dropdown. This should not happen

I was not able to replicated them, but accordion has been removed. Please see  https://github.com/openwisp/openwisp-notifications/pull/57#pullrequestreview-442624134

> * If I'm on show unread only tab and I click on mark all read, it shows me all notifications after that without switching the tab. Either switch the tab to show all or don't show read notifications.

Fixed, the tab is switched now to show all notifications. :sweat: 

> * I see you have added css and js for loader, but I can't see any loader in the notifications, I thinks it's not working properly

I have changed the implementation of showing and hiding loader. It should be visible now. :smile: 

> * Current background colors for read and unread notification are not very intuitive, I think we can keep white background for read notifications and give background color to unread notifications only. 

Sounds great, I have done that.

pandafy commented:
Closing this PR in favor of #57. The development will proceed along with web sockets in #57
